SDWebImage/SDMemoryCache.m

```diff
@@ -140,6 +149,18 @@ - (void)removeAllObjects {
     [self.weakCache removeAllObjects];
     SD_UNLOCK(self.weakCacheLock);
 }
+
+#pragma mark - Helper
++ (NSUInteger)memoryCacheCostForImage:(nonnull UIImage *)image {
+    NSUInteger cost;
+    if ([image conformsToProtocol:@protocol(SDAnimatedImage)] && [image respondsToSelector:@selector(animatedImageMemoryCost)]) {
```

Comment:
This is the start of wrong way. 
* By adding `if...else....` to judge. From the decoupling and independence of modules, `SDMemoryCache` should not know what `SDAnimatedImage` is. In other words, what if user define a subclass `UserCustomImage` of `UIImage` and has property occupy large memory? add another `if` statement to judge? And for `FLAnimatedImage`, we set it as an associate object to `UIImage`, so we also need to add a `if` statement? this is much like `SVN`, server controls many things. 
* We add a property of `animatedImageMemoryCost` for `SDAnimatedImage`, by that analogy, we also need add another property to return memory cost if user define a custom `UIImage` which can't use default implementation?
* `SDMemoryCache` stores `UIImage` instance, the cost is the memory cost of `UIImage` instance, and we can't only take rendering bytes into account, for example, `SDAnimatedImage` has a `coder` property, is retains the data of `GIF` image. 

`SDMemoryCache` stores `UIImage`, so I like my category implementation,  `SDMemoryCache` only need to call `sd_memoryCost ` of any `UIImage`, and any `UIImage` can override getter method or just call setter to return memory cost.

Reply:
> we also need add another property to return memory cost if user define a custom UIImage.
This is not a valid case. The image your provide to communicate with our framework (`SDWebImage`), it's designed to use subclass. See why we create [Animated Image Class Customization](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#customization).

Reply:
> the cost is the memory cost of UIImage instance, and we can't only take rendering bytes into account,

We care about the most important thing of performance, but not every object you attach. The compressed image data, for example, 1000*1000 JPEG image, contains only less than `100KB`, but the bytes the bitmap buffer (1000*1000*3 bits) occupy is `2.86MB`.

And this is open, which means if you really need your own custom class (Typically I don't think people do this. It's designed to be compatible for YYImage/FLImage), you can add the cost of the compressed image data

Reply:
> SDMemoryCache only need to call sd_memoryCost of any UIImag

I told that it's a more common, but not what SDWebImage our framework care about. You can attach any object using in you App or third-party framework. It's OK. But our solution for animation, base on the subclass. Why subclass is choosed ? Because the category property will be lost during many image process.

We have Force Decode feature, which will draw your image bitmap context and create a new image instance. You loss all the property.

We have [Image Transformer](https://github.com/SDWebImage/SDWebImage/wiki/Advanced-Usage#image-transformer-50). Which will do image transforming using CG or vImage, and create new UIImage instance. You custom catarogy property will lose as well.

But if we use subclass and protocol, we can add a check. It's good to see this code inside your image process code.

```
if ([image confromToProtocol:@protocol(SDAnimatedImage)]) {
    // ....
}
```

But it's bad to see this instead.

```
if (image.sd_memoryCost != 0) {
    // ....  
}
```

Reply:
`sd_memoryCost` would not equal to 0, as long as any `UIImage` is a valid image. We don't need to care any transformed image or any thing , we only need to get the memory cost before image be stored into cache, we need cached images's memory cost.

```
- (NSUInteger)sd_memoryCost {
     NSNumber *memoryCost = objc_getAssociatedObject(self, _cmd);
     if (memoryCost == nil) { // calculate default memory cost
         memoryCost = @(SDCacheCostForImage(self));
     }
     return [memoryCost unsignedIntegerValue];
 }
```

Reply:
My point is to reply why this `sd_memoryCost` this category read-write property is not good idea. We should use subclass and protocol method.